PSReadLine/PlatformWindows.cs

```diff
@@ -1015,4 +1015,125 @@ private static void TerminateStragglers()
             }
         }
     }
+
+    private static readonly object _terminalOwnerThreadHandleLock = new();
+    private static IntPtr? _terminalOwnerThreadHandle;
+
+    /// <remarks>
+    /// <para>
+    ///     This method helps to find the thread id that owns the console window of the current process. The logic is
+    ///     that at the moment we are called, most likely a current foreground window owner will be the terminal window
+    ///     owner as well. We verify that by going up the process tree and if we find the foreground process owner among
+    ///     the current process' parent hierarchy, then we consider we found it.
+    /// </para>
+    /// <para>
+    ///     The terminal process is not always the direct parent of the current process, but may be higher in the
+    ///     process tree in case PowerShell is a child of some other console process.
+    /// </para>
+    /// <para>
+    ///     The result is cached for performance reasons. To avoid thread id reuse, we keep the thread handle as well.
+    /// </para>
+    /// </remarks>
+    private static uint? GetTerminalOwnerThreadId()
+    {
+        lock (_terminalOwnerThreadHandleLock)
+        {
+            if (_terminalOwnerThreadHandle.HasValue)
+            {
+                // Already cached, validate the cache and return.
+                uint tid = GetThreadId(_terminalOwnerThreadHandle.Value);
+                if (tid == 0)
+                {
+                    // Failed to get thread handle, let's free the resources.
+                    CloseHandle(_terminalOwnerThreadHandle.Value);
+                    _terminalOwnerThreadHandle = null;
+                    return null;
+                }
+
+                return tid;
+            }
+
+            IntPtr foregroundWindow = GetForegroundWindow();
+            if (foregroundWindow != IntPtr.Zero)
+            {
+                uint tid = GetWindowThreadProcessId(foregroundWindow, out var terminalOwnerPid);
+                if (tid == 0)
+                {
+                    return null;
+                }
+
+                // We've got the process, but is it a correct process? Verify using the process parent hierarchy.
+                // Define a limit not get stuck in case processed form a loop (possible in case pid reuse).
+                const int iterationLimit = 20;
+
+                PROCESS_BASIC_INFORMATION pbi = new();
+                Process process = Process.GetCurrentProcess();
+                int activeProcessId = process.Id;
+                for (int i = 0; i < iterationLimit; ++i)
+                {
+                    int status = NtQueryInformationProcess(process.Handle, 0, out pbi, Marshal.SizeOf(pbi), out _);
+                    activeProcessId = pbi.InheritedFromUniqueProcessId.ToInt32();
```

Comment:
We already have the method `GetParentPid(Process)` (in this same file). Please directly use that helper method.